### PR TITLE
Port LBFGS from Lua optim

### LINKED ISSUE #483: Add L-BFGS
For NeuralStyle and DeepDream.

### PR DESCRIPTION
**Pull request by apaszke**
fixes #483

### CODE REVIEW

File: torch/optim/lbfgs.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    .. warning::
+        This optimizer doesn't support per-parameter options and parameter
+        groups (there can be only one).
+
+    .. warning::
+        Right now all parameters have to be on a single device. This will be
+        improved in the future.
+
+    .. note::
+        This is a very memory intensive optimizer (it requires additional
```
Comment (soumith) [inline torch/optim/lbfgs.py]: because of _gather_flat_grad, the memory required is more

File: torch/optim/lbfgs.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        assert offset == self._numel()
+
+    def step(self, closure):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model
+                and returns the loss.
+        """
+        # TODO: backward after closure?
```
Comment (soumith) [inline torch/optim/lbfgs.py]: i dont understand this TODO, can you expand

File: torch/optim/lbfgs.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                num_old = len(old_dirs)
+
+                if 'ro' not in state:
+                    state['ro'] = [None] * history_size
+                    state['al'] = [None] * history_size
+                ro = state['ro']
+                al = state['al']
+
+                for i in range(num_old):
+                    ro[i] = 1 / old_stps[i].dot(old_dirs[i])
```
Comment (soumith) [inline torch/optim/lbfgs.py]: can you make explicit all divisions on whether they are integer or floating.
here, do `1. /`

File: torch/optim/lbfgs.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+                # iteration in L-BFGS loop collapsed to use just one buffer
+                q = flat_grad.neg()
+                for i in range(num_old-1, -1, -1):
```
Comment (soumith) [inline torch/optim/lbfgs.py]: end number is `-1` or `1`?

File: torch/optim/lbfgs.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            # directional derivative
+            gtd = flat_grad.dot(d)  # g * d
+
+            # check that progress can be made along that direction
+            if gtd > -tolerance_change:
+                break
+
+            # reset initial guess for step size
+            if state['n_iter'] == 1:
+                t = min(1, 1 / abs_grad_sum) * lr
```
Comment (soumith) [inline torch/optim/lbfgs.py]: same comment about divisions

File: torch/optim/lbfgs.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            else:
+                t = lr
+
+            # optional line search: user function
+            ls_func_evals = 0
+            if line_search_fn is not None:
+                # perform line search, using user function
+                raise RuntimeError("line search function is not supported yet")
```
Comment (soumith) [inline torch/optim/lbfgs.py]: write a note or warning about this in the documentation. i didn't see any.

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/optim/lbfgs.py (hunk 4 above)
Comment (apaszke) [inline torch/optim/lbfgs.py]: yeah, because loops are exclusive 🙂 looks great, doesn't it?

File: torch/optim/lbfgs.py (hunk 6 above)
Comment (apaszke) [inline torch/optim/lbfgs.py]: I didn't document the line search argument, so I think that's enough

File: torch/optim/lbfgs.py (hunk 3 above)
Comment (apaszke) [inline torch/optim/lbfgs.py]: All tensors are always floating point, so they return floats, but I can do it just for clarity

File: torch/optim/lbfgs.py (hunk 2 above)
Comment (apaszke) [inline torch/optim/lbfgs.py]: I didn't remember if it's the responsibility of a closure, or the optimizer to call `.backward()`. I'll just remove it.

File: torch/optim/lbfgs.py (hunk 1 above)
Comment (apaszke) [inline torch/optim/lbfgs.py]: You mean it's a +1?

**Review by soumith — APPROVED** [review]

**Pull request merged**